### ISSUE TITLE
Repo: add testing around `TypeVariable`s/`IndexedAccessType`s and `getConstraintInfo`

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10606]
### Suggestion

See https://github.com/typescript-eslint/typescript-eslint/pull/10602

Also document behavior of `IndexedAccessType`s in the JSDoc and update the `getConstraintInfo` API/type definitions if necessary.

**Comment by ronami** [comment]
+1, I wasn't sure if I should mention you on the PR. Optimally, I think `getConstraintInfo` should ensure the calling code handles these. Some cases where this seems relevant:

I think this is the `IndexedAccessType` ([see for more variations](https://github.com/microsoft/TypeScript/blob/21c1a61b49082915f93e3327dad0d73205cf4273/src/compiler/types.ts#L6845-L6855)):

```ts
function test<T>(arg: T, key: keyof T) {
    arg[key];
}
```

Unconstrained type parameter in union:

```ts
function test<T>(arg: T | []) {
  arg;
}
```

I've looked into some of the existing rules that specifically deal with unconstrained type parameters and got some false positives for these cases ([playground link](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=PTAEG0EMHdISwC4FoEAsCmA7SAjANugLoAUqCCADgM4BcICAnhelQMYBOcFyLecmCAHRwA9sHYBXAlWAx4yNFlwEAlAChIVBplagAZhJ0JRmUAhYIAPABUAfMUjsA5jVDWANKADW6Bq58MInpungDKruCEoAA%2BbiqgAN5qoKAgoKwi7OzorAh4DKCYIgig2RSZ5gAmyaByiLXOANw1aWUV6JWg6ABuWGaoIhJOqKD1VANSlZgA5AgA-DV1JY5O4AGEzSmt6OXsVV29pmiDw6Ml44N4U7MLKUugoc0AvmpAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOiiJo0APbRI4MAF8QSoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):

```
// [await-thenable](https://typescript-eslint.io/rules/await-thenable)
async function test<T>(arg: T, key: keyof T, S: [] | T) {
  // correctly not reported
  await arg;
  // reported even though it shouldn't?
  await arg[key];
  // reported even though it shouldn't?
  await S;
}
```

**Comment by kirkwaiblinger** [comment]
I've reached out for advice in the TS discord here: https://discord.com/channels/508357248330760243/1324909031092916225
